### PR TITLE
server: skip zone check for PERHOST iso during attachIso

### LINKED ISSUE #5265: Mounting ISO with VMware tools fails
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
UI, ACS
~~~

##### CLOUDSTACK VERSION

~~~
4.15.1
~~~

##### SUMMARY

Mounting ISO with VMware tools fails

##### STEPS TO REPRODUCE

Screenshot 2021-08-02 110620
Screenshot 2021-08-02 110633
Screenshot 2021-08-02 110657
~~~

2021-08-02 10:44:25,491 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (qtp1136768342-18:ctx-9677a095 ctx-09d9b967) (logid:0f224c14) submit async job-9589, details: AsyncJobVO {id:9589, userId: 49, accountId: 552, instanceType: None, instanceId: null, cmd: org.apache.cloudstack.api.command.user.iso.AttachIsoCmd, cmdInfo: {"virtualmachineid":"57e879b0-00d6-406f-b7a8-b5d2d4f95a37","response":"json","ctxUserId":"49","httpmethod":"GET","ctxStartEventId":"27283","id":"ac497355-7cf2-4f29-bb84-b580e50a3d28","ctxDetails":"{\"interface com.cloud.template.VirtualMachineTemplate\":\"ac497355-7cf2-4f29-bb84-b580e50a3d28\",\"interface com.cloud.vm.VirtualMachine\":\"57e879b0-00d6-406f-b7a8-b5d2d4f95a37\"}","ctxAccountId":"552","uuid":"ac497355-7cf2-4f29-bb84-b580e50a3d28","cmdEventType":"ISO.ATTACH","_":"1627890226629"}, cmdVersion: 0, status: IN_PROGRESS, processStatus: 0, resultCode: 0, result: null, initMsid: 345052203582, completeMsid: null, lastUpdated: null, lastPolled: null, created: null, removed: null}
2021-08-02 10:44:25,492 INFO  [o.a.c.f.j.i.AsyncJobMonitor] (API-Job-Executor-123:ctx-225ea24e job-9589) (logid:742a351d) Add job-9589 into job monitoring
2021-08-02 10:44:25,494 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-123:ctx-225ea24e job-9589) (logid:ff46da57) Executing AsyncJobVO {id:9589, userId: 49, accountId: 552, instanceType: None, instanceId: null, cmd: org.apache.cloudstack.api.command.user.iso.AttachIsoCmd, cmdInfo: {"virtualmachineid":"57e879b0-00d6-406f-b7a8-b5d2d4f95a37","response":"json","ctxUserId":"49","httpmethod":"GET","ctxStartEventId":"27283","id":"ac497355-7cf2-4f29-bb84-b580e50a3d28","ctxDetails":"{\"interface com.cloud.template.VirtualMachineTemplate\":\"ac497355-7cf2-4f29-bb84-b580e50a3d28\",\"interface com.cloud.vm.VirtualMachine\":\"57e879b0-00d6-406f-b7a8-b5d2d4f95a37\"}","ctxAccountId":"552","uuid":"ac497355-7cf2-4f29-bb84-b580e50a3d28","cmdEventType":"ISO.ATTACH","_":"1627890226629"}, cmdVersion: 0, status: IN_PROGRESS, processStatus: 0, resultCode: 0, result: null, initMsid: 345052203582, completeMsid: null, lastUpdated: null, lastPolled: null, created: null, removed: null}
2021-08-02 10:44:25,510 DEBUG [o.a.c.f.j.i.AsyncJobManagerImpl] (API-Job-Executor-123:ctx-225ea24e job-9589) (logid:ff46da57) Complete async job-9589, jobStatus: FAILED, resultCode: 530, result: org.apache.cloudstack.api.response.ExceptionResponse/null/{"uuidList":[],"errorcode":"431","errortext":"ISO is not available in the zone the VM is in."}

VMware tools iso can be attached to instance directly from vCenter.
~~~

##### EXPECTED RESULTS

~~~
Vmware tools iso attached to instance
~~~

##### ACTUAL RESULTS

~~~
Vmware tools iso is not attached to instance
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Hypervisor tools ISO - vmware-toools.iso, xs-tools.iso are marked as PERHOST in DB. They are active but not downloaded to the secondary storages and hence no template-zone entry.
Skips the template-zone check for such templates.

Fixes #5265

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### Screenshots (if appropriate):
Error on attach ISO:
Screenshot from 2021-08-03 13-17-17

### How Has This Been Tested?

Using cmk with an env (VMware advanced zone)

Before fix:
```
(localcloud) SBCM5> > list isos id=8fd33e21-d145-4b23-91f0-a0cecc9429fa isofilter=self
{
  "count": 1,
  "iso": [
    {
      "account": "system",
      "bits": 64,
      "bootable": false,
      "crossZones": false,
      "directdownload": false,
      "displaytext": "VMware Tools Installer ISO",
      "domain": "ROOT",
      "domainid": "c4a349e5-f487-11eb-ac74-1e0019000402",
      "id": "8fd33e21-d145-4b23-91f0-a0cecc9429fa",
      "isdynamicallyscalable": false,
      "isextractable": false,
      "isfeatured": true,
      "ispublic": true,
      "isready": true,
      "name": "vmware-tools.iso",
      "ostypeid": "c4aaad30-f487-11eb-ac74-1e0019000402",
      "ostypename": "CentOS 4.5 (32-bit)",
      "passwordenabled": false,
      "tags": []
    }
  ]
}
(localcloud) SBCM5> > list virtualmachines id=8dbc5889-cdce-49f7-a23b-a3e862c5316d
{
  "count": 1,
  "virtualmachine": [
    {
      "account": "admin",
      "affinitygroup": [],
      "cpunumber": 1,
      "cpuspeed": 500,
      "cpuused": "0%",
      "created": "2021-08-03T18:52:00+0000",
      "details": {
        "cpuOvercommitRatio": "2.0",
        "dataDiskController": "osdefault",
        "memoryOvercommitRatio": "1.0",
        "rootDiskController": "ide"
      },
      "diskioread": 62,
      "diskiowrite": 46,
      "diskkbsread": 2512,
      "diskkbswrite": 349,
      "diskofferingid": "a93176e0-6392-4e39-aa30-c798ca33cae1",
      "diskofferingname": "Large",
      "displayname": "t1",
      "displayvm": true,
      "domain": "ROOT",
      "domainid": "c4a349e5-f487-11eb-ac74-1e0019000402",
      "guestosid": "c4ac616b-f487-11eb-ac74-1e0019000402",
      "haenable": false,
      "hostid": "8a14a9f9-ebbc-4eaa-9f57-940a3210e913",
      "hostname": "10.0.32.172",
      "hypervisor": "VMware",
      "id": "8dbc5889-cdce-49f7-a23b-a3e862c5316d",
      "instancename": "i-2-3-VM",
      "isdynamicallyscalable": false,
      "memory": 512,
      "memoryintfreekbs": 393216,
      "memorykbs": 524288,
      "memorytargetkbs": 524288,
      "name": "t1",
      "networkkbsread": 0,
      "networkkbswrite": 0,
      "nic": [
        {
          "broadcasturi": "vlan://1530",
          "extradhcpoption": [],
          "gateway": "10.1.1.1",
          "id": "eabbdb90-aa81-4e8a-8da3-02de530b3df0",
          "ipaddress": "10.1.1.192",
          "isdefault": true,
          "isolationuri": "vlan://1530",
          "macaddress": "02:00:6c:ea:00:01",
          "netmask": "255.255.255.0",
          "networkid": "33dd6451-bc69-4587-ae9e-e2add4263d9c",
          "networkname": "net1",
          "secondaryip": [],
          "traffictype": "Guest",
          "type": "Isolated"
        }
      ],
      "osdisplayname": "CentOS 5.3 (64-bit)",
      "ostypeid": "c4ac616b-f487-11eb-ac74-1e0019000402",
      "passwordenabled": false,
      "rootdeviceid": 0,
      "rootdevicetype": "ROOT",
      "securitygroup": [],
      "serviceofferingid": "01410b2a-4471-4a7c-9e2d-84fd9ebff363",
      "serviceofferingname": "Small Instance",
      "state": "Running",
      "tags": [],
      "templatedisplaytext": "CentOS 5.3(64-bit) no GUI (vSphere)",
      "templateid": "c4a8b7d0-f487-11eb-ac74-1e0019000402",
      "templatename": "CentOS 5.3(64-bit) no GUI (vSphere)",
      "userid": "dd534934-f487-11eb-ac74-1e0019000402",
      "username": "admin",
      "zoneid": "358d9914-5590-4b6a-b615-ae3b83af52fa",
      "zonename": "ref-trl-1450-v-M7-abhishek-kumar"
    }
  ]
}
(localcloud) SBCM5> > attach iso virtualmachineid=8dbc5889-cdce-49f7-a23b-a3e862c5316d id="8fd33e21-d145-4b23-91f0-a0cecc9429fa"
{
  "accountid": "dd522416-f487-11eb-ac74-1e0019000402",
  "cmd": "org.apache.cloudstack.api.command.admin.iso.AttachIsoCmdByAdmin",
  "completed": "2021-08-03T18:56:19+0000",
  "created": "2021-08-03T18:56:18+0000",
  "jobid": "a776dfb2-130b-40a9-a8d5-1d0e1b9cf83b",
  "jobprocstatus": 0,
  "jobresult": {
    "errorcode": 431,
    "errortext": "ISO is not available in the zone the VM is in."
  },
  "jobresultcode": 530,
  "jobresulttype": "object",
  "jobstatus": 2,
  "userid": "dd534934-f487-11eb-ac74-1e0019000402"
}
� Error: async API failed for job a776dfb2-130b-40a9-a8d5-1d0e1b9cf83b
```

After fix:
```
(localcloud) SBCM5> > attach iso virtualmachineid=8dbc5889-cdce-49f7-a23b-a3e862c5316d id="8fd33e21-d145-4b23-91f0-a0cecc9429fa"
{
  "virtualmachine": {
    "account": "admin",
    "affinitygroup": [],
    "cpunumber": 1,
    "cpuspeed": 500,
    "cpuused": "5%",
    "created": "2021-08-03T18:52:00+0000",
    "details": {
      "cpuOvercommitRatio": "2.0",
      "dataDiskController": "osdefault",
      "memoryOvercommitRatio": "1.0",
      "rootDiskController": "ide"
    },
    "diskioread": 0,
    "diskiowrite": 0,
    "diskkbsread": 0,
    "diskkbswrite": 0,
    "diskofferingid": "a93176e0-6392-4e39-aa30-c798ca33cae1",
    "diskofferingname": "Large",
    "displayname": "t1",
    "displayvm": true,
    "domain": "ROOT",
    "domainid": "c4a349e5-f487-11eb-ac74-1e0019000402",
    "guestosid": "c4ac616b-f487-11eb-ac74-1e0019000402",
    "haenable": false,
    "hostid": "8a14a9f9-ebbc-4eaa-9f57-940a3210e913",
    "hostname": "10.0.32.172",
    "hypervisor": "VMware",
    "id": "8dbc5889-cdce-49f7-a23b-a3e862c5316d",
    "instancename": "i-2-3-VM",
    "isdynamicallyscalable": false,
    "isodisplaytext": "VMware Tools Installer ISO",
    "isoid": "8fd33e21-d145-4b23-91f0-a0cecc9429fa",
    "isoname": "vmware-tools.iso",
    "memory": 512,
    "memoryintfreekbs": 77824,
    "memorykbs": 524288,
    "memorytargetkbs": 524288,
    "name": "t1",
    "networkkbsread": 0,
    "networkkbswrite": 0,
    "nic": [
      {
        "broadcasturi": "vlan://1530",
        "extradhcpoption": [],
        "gateway": "10.1.1.1",
        "id": "eabbdb90-aa81-4e8a-8da3-02de530b3df0",
        "ipaddress": "10.1.1.192",
        "isdefault": true,
        "isolationuri": "vlan://1530",
        "macaddress": "02:00:6c:ea:00:01",
        "netmask": "255.255.255.0",
        "networkid": "33dd6451-bc69-4587-ae9e-e2add4263d9c",
        "networkname": "net1",
        "secondaryip": [],
        "traffictype": "Guest",
        "type": "Isolated"
      }
    ],
    "osdisplayname": "CentOS 5.3 (64-bit)",
    "ostypeid": "c4ac616b-f487-11eb-ac74-1e0019000402",
    "passwordenabled": false,
    "rootdeviceid": 0,
    "rootdevicetype": "ROOT",
    "securitygroup": [],
    "serviceofferingid": "01410b2a-4471-4a7c-9e2d-84fd9ebff363",
    "serviceofferingname": "Small Instance",
    "state": "Running",
    "tags": [],
    "templatedisplaytext": "CentOS 5.3(64-bit) no GUI (vSphere)",
    "templateid": "c4a8b7d0-f487-11eb-ac74-1e0019000402",
    "templatename": "CentOS 5.3(64-bit) no GUI (vSphere)",
    "userid": "dd534934-f487-11eb-ac74-1e0019000402",
    "username": "admin",
    "zoneid": "358d9914-5590-4b6a-b615-ae3b83af52fa",
    "zonename": "ref-trl-1450-v-M7-abhishek-kumar"
  }
}
```

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 1]
```diff
@@ -1204,10 +1207,11 @@ public boolean attachIso(long isoId, long vmId, boolean forced) {
             throw new InvalidParameterValueException("Unable to find an ISO with id " + isoId);
         }
 
-        long dcId = vm.getDataCenterId();
-        VMTemplateZoneVO exists = _tmpltZoneDao.findByZoneTemplate(dcId, isoId);
-        if (null == exists) {
-            throw new InvalidParameterValueException("ISO is not available in the zone the VM is in.");
+        if (TemplateType.PERHOST.equals(iso.getTemplateType())) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:1210]: if it was PERHOST we were not going to check, right? looks reversed here.

Comment (shwstppr) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @DaanHoogland thanks. Yeah made mistake

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM.

**Comment by shwstppr** [comment]
cmk output added for tests using VMware env attaching vmware-tools ISO to a VM

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by weizhouapache** [comment]
tested ok.
I merged this pr into #5258 , and confirmed the ISO vmware-tools.iso can be attached to vm.

**Review by weizhouapache — APPROVED** [review]

**Review by sureshanaparti — APPROVED** [review]
clgtm

**Comment by sureshanaparti** [comment]
@blueorangutan test matrix

**Review by davidjumani — APPROVED** [review]
Verified, able to mount `vmware-tools.iso` and `xs-tools.iso`

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 2]
```diff
@@ -300,6 +300,9 @@
     @Inject
     private EndPointSelector selector;
 
+    private final static String VMWARE_TOOLS_ISO = "vmware-tools.iso";
```
Comment (nvazquez) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:303]: Just an improvement suggestion: I've noticed these ISO names are hardcoded on the Vmware and Xen plugins codebase. What about moving `VMWARE_TOOLS_ISO` and `XS_TOOLS_ISO` to the `TemplateManager` interface and replace the hardcoded ISO names for these values on the `plugins/hypervisor/vmware` and `plugins/hypervisor/xenserver` modules?

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java (hunk 2 above)
Comment (shwstppr) [reply server/src/main/java/com/cloud/template/TemplateManagerImpl.java]: @nvazquez done

**Review by nvazquez — APPROVED** [review]
Code LGTM

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@blueorangutan test centos7 xs71

**Comment by shwstppr** [comment]
@blueorangutan test centos7 xenserver-71

**Pull request merged**